site/blocks/StickyNav.tsx

```diff
@@ -0,0 +1,202 @@
+                currentHeadingIndex = i
+            }
+        }
+
+        const current = this.ulRef.current
+        if (current) {
+            const parent = current.parentElement
+            if (parent) {
+                const hasScrollbar = parent.scrollWidth < current.scrollWidth
+                if (hasScrollbar) {
+                    const listItem = current.children[
+                        currentHeadingIndex
+                    ] as HTMLElement
+                    const containerPadding = 32
+                    current.scrollTo({
+                        behavior: "smooth",
+                        left: listItem.offsetLeft - containerPadding,
+                    })
+                }
+            }
+        }
+
+        this.setState({
+            currentHeadingIndex,
+        })
+    }
+
+    setHeadingPositions() {
+        const headingPositions: { id: string; yPosition: number }[] = []
+
+        this.state.links.forEach(({ target }, i) => {
+            const element = document.querySelector<HTMLElement>(target)
+            if (element) {
+                const { y } = getTotalOffset(element)
+                headingPositions.push({ id: target, yPosition: y })
+            } else {
+                // Because of filterValidLinks, this should never happen
+                // But if it does, pretend the heading exists
+                // 1 pixel after the preceding heading
+                headingPositions.push({
+                    id: target,
+                    yPosition: headingPositions[i - 1].yPosition + 1,
+                })
+            }
+        })
+
+        this.setState(
+            {
+                headingPositions,
+            },
+            () => {
+                this.setCurrentHeading()
+            }
+        )
+    }
+
+    componentDidMount() {
+        this.filterValidLinks()
+        // Web fonts change the position of elements
+        // so we set the active heading only once we know where everything is.
+        // The alternative is to call setHeadingPositions as soon as the component mounts,
+        // but this can make currentHeadingIndex jump around.
+        window.addEventListener("load", this.handleResize)
+        window.addEventListener("scroll", this.handleScroll)
+        window.addEventListener("resize", this.handleResize)
+    }
+
+    componentWillUnmount() {
+        window.removeEventListener("scroll", this.handleScroll)
+        window.removeEventListener("resize", this.handleResize)
+        window.removeEventListener("load", this.handleResize)
+    }
+
+    render() {
+        if (!this.state.links) return null
+        return (
+            <>
+                <style>
+                    {/* add scroll-margin-top to all elements with an ID */}
+                    {`
+                    [id] {
+                         scroll-margin-top:${
+                             STICKY_NAV_HEIGHT + HEADING_SCROLL_MARGIN
+                         }px;
```

Comment:
Hm, I'm a little confused now as to why we need this.
From what I read, it seems that `scroll-margin-top` only applies if `scroll-snap-type` is set. Are you certain that setting it here actually makes a difference?

Reply:
It makes it so that the anchor links navigate to a point above each heading, so that the nav bar doesn't obscure the heading. `scroll-snap-type` doesn't need to be set for this, afaik. 🙂 

https://user-images.githubusercontent.com/11844404/182887019-9b537a16-8404-4e58-8ab0-65e01d952fe5.mov

(I checked with Chrome as well)

Reply:
Ah, that makes sense. And I now realize that I never actually tried using the nav this way while reviewing :)